bundles/org.openhab.voice.rustpotterks/src/main/java/org/openhab/voice/rustpotterks/internal/RustpotterKSService.java

```diff
@@ -0,0 +1,245 @@
+import io.github.givimad.rustpotter_java.VadMode;
+
+/**
+ * The {@link RustpotterKSService} is responsible for creating things and thing
+ * handlers.
+ *
+ * @author Miguel Álvarez - Initial contribution
+ */
+@NonNullByDefault
+@Component(configurationPid = SERVICE_PID, property = Constants.SERVICE_PID + "=" + SERVICE_PID)
+@ConfigurableService(category = SERVICE_CATEGORY, label = SERVICE_NAME
+        + " Keyword Spotter", description_uri = SERVICE_CATEGORY + ":" + SERVICE_ID)
+public class RustpotterKSService implements KSService {
+    private static final String RUSTPOTTER_FOLDER = Path.of(OpenHAB.getUserDataFolder(), "rustpotter").toString();
+    private final Logger logger = LoggerFactory.getLogger(RustpotterKSService.class);
+    private final ScheduledExecutorService executor = ThreadPoolManager.getScheduledPool("OH-voice-porcupineks");
+    private RustpotterKSConfiguration config = new RustpotterKSConfiguration();
+    static {
+        Logger logger = LoggerFactory.getLogger(RustpotterKSService.class);
+        File directory = new File(RUSTPOTTER_FOLDER);
+        if (!directory.exists()) {
+            if (directory.mkdir()) {
+                logger.info("rustpotter dir created {}", RUSTPOTTER_FOLDER);
+            }
+        }
+    }
+
+    @Activate
+    protected void activate(ComponentContext componentContext, Map<String, Object> config) {
+        modified(config);
+    }
+
+    @Modified
+    protected void modified(Map<String, Object> config) {
+        this.config = new Configuration(config).as(RustpotterKSConfiguration.class);
+    }
+
+    @Override
+    public String getId() {
+        return SERVICE_ID;
+    }
+
+    @Override
+    public String getLabel(@Nullable Locale locale) {
+        return SERVICE_NAME;
+    }
+
+    @Override
+    public Set<Locale> getSupportedLocales() {
+        return Set.of(Locale.ENGLISH, new Locale("es"), Locale.FRENCH, Locale.GERMAN);
+    }
+
+    @Override
+    public Set<AudioFormat> getSupportedFormats() {
+        return Set
+                .of(new AudioFormat(AudioFormat.CONTAINER_WAVE, AudioFormat.CODEC_PCM_SIGNED, null, null, null, null));
+    }
+
+    @Override
+    public KSServiceHandle spot(KSListener ksListener, AudioStream audioStream, Locale locale, String keyword)
+            throws KSException {
+        logger.debug("Loading library");
+        try {
+            RustpotterJava.loadLibrary();
+        } catch (IOException e) {
+            throw new KSException("Unable to load rustpotter lib: " + e.getMessage());
+        }
+        var audioFormat = audioStream.getFormat();
+        var frequency = Objects.requireNonNull(audioFormat.getFrequency());
+        var bitDepth = Objects.requireNonNull(audioFormat.getBitDepth());
+        var channels = Objects.requireNonNull(audioFormat.getChannels());
+        var endianness = Objects.requireNonNull(audioFormat.isBigEndian()) ? Endianness.BIG : Endianness.LITTLE;
```

Comment:
As far I understand yes, the interface is just nullable to allow describing the supported formats, but the function that checks the format support ensures these are not null.

Reply:
Can we point me the function you are talking about please ?

Reply:
Sure, it's the  'getBestMatch' method from the VoiceManagerImpl class on the core.
It's calling 'getPreferredFormat' which is the one it's validating this.
At it make sense for me because you can not process a wav stream without all that info.

Reply:
I am not sure at all the method getBestMatch guarantees that none of these data will be non null. As an example there is no check on the number of channels.
So, instead of `Objects.requireNonNull` that could trigger a NullPointerException, I would suggest, in case you absolutely need that these values to be set, to rather trigger the interpretation exception.

Reply:
I think the method initialize the channels to 1 in case they are null on both sides, but still it makes sense for me to do the change just in case.